### PR TITLE
Upgrade/improve signer for iframe triggered by fast-auth-wallet 

### PR DESCRIPTION
**Pull request by benmalcom**
- This PR improves how the fast-auth-signer is rendered in iframe, triggered by the fast-auth-wallet [in this PR](https://github.com/near/near-fastauth-wallet/pull/14).
-  Particular uses this hook `useIframeDialogConfig` to get the height of the wrapper content to be loaded into iframe and sets it as the modal/drawer height.

### CODE REVIEW

File: packages/near-fast-auth-signer/src/components/Login/Login.tsx [hunk 1]
```diff
@@ -17,27 +22,17 @@ const schema = yup.object().shape({
     .required('Please enter a valid email address'),
 });
 
+const LoginForm = styled(FormContainer)`
+  height: 400px;
+`;
+
 function Login() {
+  // Send form height to modal if in iframe
+  useIframeDialogConfig({ element: document.querySelector('#loginForm') });
```
Comment (Pessina) [inline packages/near-fast-auth-signer/src/components/Login/Login.tsx:31]: Also, can we user React Refs?

File: packages/near-fast-auth-signer/src/components/AddDevice/AddDevice.tsx [hunk 2]
```diff
@@ -78,7 +56,9 @@ const schema = yup.object().shape({
     .required('Please enter a valid email address'),
 });
 
-function SignInPage() {
+function AddDevicePage() {
+  useIframeDialogConfig({ element: document.querySelector('#addDeviceForm') });
```
[the remark this answers is not in the document]
Comment (benmalcom) [reply packages/near-fast-auth-signer/src/components/AddDevice/AddDevice.tsx:60]: Good catch

**Review by esaminu — DISMISSED** [review]
LGTM

File: packages/near-fast-auth-signer/.gitignore [hunk 3]
```diff
@@ -25,3 +25,6 @@ yarn-error.log*
```
Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:25]: Screenshot 2024-01-25 at 3 01 09 PM

On iPad mini the iFrame it's not properly rendered

Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:25]: Not sure if this is ok, but when there is a error message, the `Connect Wallet` will be almost touching the lower border of the modal. 

Screenshot 2024-01-25 at 3 25 53 PM

Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:25]: I'm not able to create an account from Safari, I'm getting the following error: http://localhost:3002/?reason=Firebase%3A+Error+%28auth%2Fnetwork-request-failed%29.


https://github.com/near/fast-auth-signer/assets/22358004/a4d36070-fda9-4166-9530-d7fcbed33eb4

**Review by Pessina — CHANGES_REQUESTED** [review]

File: packages/near-fast-auth-signer/.gitignore [hunk 4]
```diff
@@ -25,3 +25,6 @@ yarn-error.log*
 
 # Sentry Config File
```
Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:26]: Weird behavior with passkeys: 

1 - When I click SignIn the passkeys selector it's prompted
2 - I select a passkey
3 - A different email will appear on the input

https://drive.google.com/file/d/1RSNgGsm15mNNWdSx8RHSwRlZpc2giZoj/view?usp=sharing

Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:26]: Weird behavior with passkeys: 

1 - When I click SignIn the passkeys selector it's prompted
2 - I select a passkey
3 - A different email will appear on the input
4 - Click login
5 - User will be logged in with the email selected on passkey

https://drive.google.com/file/d/1HY-5EK0bk1SciU72sHFAf1pk7mSmYoJI/view?usp=sharing

The passkey I choose here it's from Apple Profile, not Chrome. Don't know if it makes difference

Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:26]: Verify email it's send to a device where the user recently logged in

1 - On Firefox create an account
2 - After login, logout and click SignIn
3 - Before the email it's autofilled, paste the email copied from MinuteMail
4 - Click Login quickly
5 - You will end up on VerifyEmail page even thought the device was recently added

https://drive.google.com/file/d/1N8Pg-M77sZZGVH4Nyki19dUvmHznL3_H/view?usp=sharing

**Review by Pessina — CHANGES_REQUESTED** [review]

File: packages/near-fast-auth-signer/.gitignore (hunk 3 above)
Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:25]: We probably need a re-design of this too 
Screenshot 2024-02-06 at 9 08 23 PM

**Review by Pessina — CHANGES_REQUESTED** [review]

File: packages/near-fast-auth-signer/.gitignore (hunk 4 above)
Comment (benmalcom) [reply packages/near-fast-auth-signer/.gitignore:26]: @esaminu

Comment (benmalcom) [reply packages/near-fast-auth-signer/.gitignore:26]: @esaminu

Comment (benmalcom) [reply packages/near-fast-auth-signer/.gitignore:26]: @esaminu

**Comment by esaminu** [comment]
@Pessina thanks for catching this stuff. https://github.com/near/fast-auth-signer/pull/155#discussion_r1477120744 , https://github.com/near/fast-auth-signer/pull/155#discussion_r1477120947 and https://github.com/near/fast-auth-signer/pull/155#discussion_r1477121068 are indexer issues. You are sometimes able to login by verifying email because the mpc key is old enough to have been indexed by the indexer but you aren't able to login with passkey because the passkey's keypair addition was not indexed yet by the indexer. I'm wondering if we can mitigate this as part of https://github.com/near/fast-auth-signer/issues/161 by intercepting add key actions and using firestore as the first layer

**Comment by Pessina** [comment]
@esaminu, I don't believe this issue is related to KitWallet. I've merged the [Firebase Indexer PR](https://github.com/near/fast-auth-signer/issues/161) and created a recording where I set a breakpoint on fetchAccountIds. The accountId is correctly returned, yet the bug persists.

Recording: https://drive.google.com/file/d/1vn3GMRxJM3SjMRvTGJ8dJrPWcVTBKFRT/view?usp=sharing

Could you provide more details on how KitWallet is related to this bug? I'm not seeing the connection.

Bugs:

- The email selected via passkey differs from the email auto-filled in the input field. It might confuse users to select one email and see another filled in.
- The user is logged in with the email selected from the passkey, not the one in the email input field. This behavior is already present in the main branch, but it seems less confusing there because the passkey is prompted last, allowing the user to remember their last choice more clearly.

I propose we maintain the current behavior and consider improving the flow later if necessary:

- Remove the auto-fill feature for the email input.
- Prompt for the passkey after the user inputs an email.

But, I'm not a specialist, what do you think, @nall-near?

**Comment by esaminu** [comment]
@Pessina The issues with redirecting to verify email is the result of the accountId. For the auto fill issue it should be prefilled before it prompts for passkey in passkey supported browsers. Can you pull the latest and check this is the behavior? In non webauthn browsers there should be no auto fill

File: packages/near-fast-auth-signer/.gitignore (hunk 3 above)
Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:25]: Passkeys prompt two times if user click on Confirm button after select passkeys

https://drive.google.com/file/d/1pUpJmao7dQVpL8Bb3eBATxl46qxnu4DN/view?usp=sharing

Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:25]: Only able to login in Safari with email link (passkeys not working)

https://drive.google.com/file/d/1pbzvK7Rj5OglOMVDX1CLMB2o6nxG9G0W/view?usp=sharing

**Review by Pessina — CHANGES_REQUESTED** [review]

**Review by esaminu — APPROVED** [review]
LGTM

File: packages/near-fast-auth-signer/src/api/index.ts [hunk 5]
```diff
@@ -9,35 +10,26 @@ import {
  * Fetches the account IDs associated with a given public key.
  *
  * @param publicKey - The public key to fetch the account IDs for.
- * @param options - An object containing the following properties:
- * - returnEmpty: A boolean indicating whether to return an empty array if no account IDs are found.
  * @returns A promise that resolves to an array of account IDs.
  * @throws Will throw an error if the fetch request fails.
  */
+export const fetchAccountIds = async (publicKey: string): Promise<string[]> => {
```
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/api/index.ts:16]: BLOCKER

File: packages/near-fast-auth-signer/src/components/AddDevice/AddDevice.tsx [hunk 6]
```diff
@@ -127,196 +143,232 @@ function SignInPage() {
...
+      : null;
+    const existingDeviceLakKey = existingDevice?.publicKeys?.filter((key) => key !== publicKeyFak)[0];
+
+    // @ts-ignore
+    const oidcToken = firebaseUser?.accessToken;
+    const recoveryPk = oidcToken && (await window.fastAuthController.getUserCredential(oidcToken).catch(() => false));
+    const allKeys = [public_key, publicKeyFak].concat(recoveryPk || []);
+    // if given lak key is already attached to webAuthN public key, no need to add it again
+    const noNeedToAddKey = existingDeviceLakKey === public_key;
+    const parsedUrl = new URL(success_url || window.location.origin + (basePath ? `/${basePath}` : ''));
```
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/components/AddDevice/AddDevice.tsx:198]: BLOCKER

File: packages/near-fast-auth-signer/src/components/AddDevice/AddDevice.tsx [hunk 7]
```diff
@@ -127,196 +143,232 @@ function SignInPage() {
...
-
-  useEffect(() => {
-    if (authenticated === 'loading') return;
+      })
+      .finally(() => setInFlight(false)); // @ts-ignore
+  }, [firebaseUser, navigate, searchParams]);
 
-    const handleAuthCallback = async () => {
+  const onSubmit = async (data: { email: string }) => {
+    if (!data.email) return;
```
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/components/AddDevice/AddDevice.tsx:283]: BLOCKER

File: packages/near-fast-auth-signer/src/components/InitializeGlobals/InitializeGlobals.tsx [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import FirestoreController from '../../lib/firestoreController';
+import { networkId } from '../../utils/config';
+
+const TEST_ACCOUNT_ID = 'harisvalj.testnet';
+window.fastAuthController = new FastAuthController({
+  accountId: TEST_ACCOUNT_ID,
+  networkId
+});
+
+function InitializeGlobals() {
```
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/components/InitializeGlobals/InitializeGlobals.tsx:13]: BLOCKER

File: packages/near-fast-auth-signer/src/components/InitializeGlobals/InitializeGlobals.tsx [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import { useEffect } from 'react';
+
+import FastAuthController from '../../lib/controller';
+import FirestoreController from '../../lib/firestoreController';
+import { networkId } from '../../utils/config';
+
+const TEST_ACCOUNT_ID = 'harisvalj.testnet';
```
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/components/InitializeGlobals/InitializeGlobals.tsx:7]: BLOCKER

File: packages/near-fast-auth-signer/src/components/Sign/Sign.tsx [hunk 10]
```diff
@@ -167,125 +177,133 @@ function Sign() {
...
+      window.parent.postMessage({ signedDelegates: '', error: errorMessage }, '*');
+      setInFlight(false);
+      return;
+    }
+    const signedTransactions = [];
+    const success_url = isUrlNotJavascriptProtocol(searchParams.get('success_url')) && searchParams.get('success_url');
+    for (let i = 0; i < transactionDetails.transactions.length; i += 1) {
+      try {
+        // eslint-disable-next-line no-await-in-loop
+        const signed = await window.fastAuthController.signDelegateAction(
```
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/components/Sign/Sign.tsx:195]: BLOCKER

File: packages/near-fast-auth-signer/src/api/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/api/index.ts:16]: BLOCKER

File: packages/near-fast-auth-signer/src/components/InitializeGlobals/InitializeGlobals.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (esaminu) [reply packages/near-fast-auth-signer/src/components/InitializeGlobals/InitializeGlobals.tsx:7]: This is on main branch so not a blocker

File: packages/near-fast-auth-signer/src/components/Sign/Sign.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (esaminu) [reply packages/near-fast-auth-signer/src/components/Sign/Sign.tsx:195]: This is current behavior on main so not a blocker

[the remark this answers is not in the document]
Comment (benmalcom) [reply packages/near-fast-auth-signer/src/components/Sign/Sign.tsx:195]: First this is not a blocker.
This is how it works currently in prod, we'll address these concerns once we have a workable solution.

**Review by esaminu — APPROVED** [review]
LGTM

File: packages/near-fast-auth-signer/src/components/Sign/Sign.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/components/Sign/Sign.tsx:195]: Agree, if it's on main already it's not a blocker. My mistake for tagging it as blocker

File: packages/near-fast-auth-signer/src/components/InitializeGlobals/InitializeGlobals.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (Pessina) [reply packages/near-fast-auth-signer/src/components/InitializeGlobals/InitializeGlobals.tsx]: Agree, if it's on main already it's not a blocker. My mistake for tagging it as blocker

File: packages/near-fast-auth-signer/.gitignore (hunk 3 above)
Comment (Pessina) [inline packages/near-fast-auth-signer/.gitignore:25]: Steps to reproduce

1 - Create account on Chrome
2 - Logout from Chrome
2 - Login on Safari
3 - Logout from Safari
4 - Try to login from Safari
5 - Passkeys won't be requested and e-mail will be sent. 

Screen recording

https://drive.google.com/file/d/1zcvuILfF-4-L3zIY3LRCjVX2th6CyOhn/view?usp=sharing

**Review by Pessina — CHANGES_REQUESTED** [review]

File: packages/near-fast-auth-signer/.gitignore (hunk 3 above)
Comment (esaminu) [reply packages/near-fast-auth-signer/.gitignore:25]: @Pessina please open a separate issue for this thanks

**Pull request merged**